### DISCUSSION TITLE
Adding support for MarkDoc

### DISCUSSION THREAD

**Discussion by choilive** [tinacms/tinacms#3335] [Ideas]
I realize this would be a challenging proposition, but has the idea of supporting MarkDoc as an alternative to MDX been floated around? 

With the intent of this project to provide a user-friendly CMS for devs and content creators alike - MarkDoc seems particularly in line with the project's goals around making content easy to edit. See: https://markdoc.dev/docs/faq#why-not-mdx

Its been getting quite a bit of traction too - I have used MarkDoc for a documentation site and quite like the format as opposed to MDX.

**Comment by jeffsee55** [comment]
> Forgive the long post - but we've had a few questions about the state of MDX support and what that really means with Tina, so wanted to get some of this written down.
> TLDR - I would love it if we could, but we might struggle to implement it because markdoc uses markdown-it and we're using remark.

This is a great question, markdoc is interesting and definitely something we _could_ support in some capacity. 

The "why not MDX" section is a good read, and the points they make in that section are similar to discussions we've had internally.

However, instead of building another syntax, we decided to lean on a subset of MDX, which essentially means: **Tina does not support any Javascript that we can't serialize**. This means we can't support things like render props, import statements or arbitrary variable declarations. This limitation was something we weren't sure would be received well, but so far we haven't had any complaints about it. And the benefits from leveraging an already robust library have allowed us to get up and running quickly. It is funny though because these limitations make it more similar to markdoc than it is to "regular" MDX. So there very well could be a path forward here.

> "Markdoc enforces a strict separation between code and content (think: docs as data)." 

I think this is a solid point and not one I hear much regarding MDX. MDX is _deeply_ integrated into your code, so much so that frameworks have decided it's easier to bring MDX support baked into their build tooling (Next, Remix, Astro off the top of my head). This coupling is a total showstopper for a CMS, we can't be involved in build tooling because the whole point of a CMS is to bring editing capabilities to editors instead of developers. For us, the solution here was to serialize the MDX tree and treat it as a plain structured content. As mentioned above, this means limiting what we support from MDX, and the lack of support for import statements means we're also decoupling content from code in much the same way as markdoc (because you register that component at render time with `<TinaMarkdown>`). So all in all, on it's surface it looks like we should be able to support markdoc with the same patterns, however...

One thing that's not ideal for us is that markdoc is built on top of [markdown-it](https://github.com/markdown-it/markdown-it). While MDX uses remark. It's too bad that they're not built on top of the same tooling because that'd probably mean we'd need an entirely different set of parsing/stringifying logic to support both. Not only that, but the thing about markdown-it is that from my brief investigation, formatting it for our rich-text editor seemed like a much larger leap than remark due to the shape of their AST. (read details for more info).

<details>
Markdown-it's AST is much more flattened out, so an opening paragraph tag is denoted with `type: 'paragraph_open'`, and everything until `type: 'paragraph_close'` needs to be reparented into a `paragraph` node for our rich-text editor.

For remark, the shape is is nested, which is closer to the shape expected by our rich-text editor (which uses [slate](https://docs.slatejs.org/)
</details>

Regardless of whether or not we can realistically add support for markdoc, one thing I would love to add support for is references to values in frontmatter. Markdoc [does this well](https://markdoc.dev/docs/frontmatter), and so does Astro's markdown support. We can do this too, but obviously it's more work for us because of the rich-text editor.

---

I should also mention markdown directives, which look and feel _very_ similar to markdoc. That may be the most realistic non-JSX solution for us to support as there's a [remark plugin](https://github.com/remarkjs/remark-directive) for it.


---

One last point, and this pertains to markdown directives too. One thing that sort of jams us up is that these syntaxes don't support _multiple_ rich-text properties. For example, the code from their docs:

```
{% callout type="check" %}
Markdoc is open-source—check out its [source](http://github.com/markdoc/markdoc) to see how it works.
{% /callout %}
```
With Tina, you would register `callout` as a template, and you'd have a `string` field with `name: "type"`, and a `rich-text` field with `name: "children"`.

But what if you want another rich text field as part of this component?

```
{% callout type="check" disclaimer="Read the [disclaimer](some-link) before use." %}
Markdoc is open-source—check out its [source](http://github.com/markdoc/markdoc) to see how it works.
{% /callout %}
```
A multiline `disclaimer` value wouldn't be parsed properly, but with MDX we can do this:

```jsx
<Callout type="check" disclaimer={<>
    Read the [disclaimer](some-link) before use.
  </>}>
  Markdoc is open-source—check out its [source](http://github.com/markdoc/markdoc) to see how it works.
</Callout>
```
That might seem like a small thing, and there are some patterns you could use to avoid this, but Tina has a responsibility to _not_ expose technical limitations to the (likely non-technical) editor working with content. And it's really nice to not have to worry about this type of thing when you're designing your schema.
